### PR TITLE
Updates to batch checks for simulation restarts

### LINKED ISSUE #2389: Simulation restart isn't terminated correctly
As @eepeterson and I were discussing restart capabilities today, I noted that the restart of a problem without editing the number of batches in the `settings.xml` file results in an unbounded simulation. I believe this is caused by the following lines in `simulation.cpp`:

https://github.com/openmc-dev/openmc/blob/6e2a9add5ef5019147e20422c380900c9b4a21a4/src/simulation.cpp#L238-L246

When a restart occurs and the initial number of batches read from the statepoint file is equal to `settings::n_max_batches` set by the `settings.xml` file, the simulation will continue indefinitely.

This probably should be caught here:

https://github.com/openmc-dev/openmc/blob/6e2a9add5ef5019147e20422c380900c9b4a21a4/src/state_point.cpp#L406-L409

I think a couple of changes are in order:

  - change the check in `simulation.cpp`  from `==` to `>=`
  - change the check in `state_point.cpp` from `>` to `>=` and update the error message appropriately

### PR DESCRIPTION
**Pull request by pshriwise**
Resolves #2389 

Additional changes:
  - change to allow `openmc.Pathlike` to be passed for the restart file in the Python API
  - "Pathlike" --> "PathLike" in docstrrings

### CODE REVIEW

**Comment by pshriwise** [comment]
A couple more updates here:

  - Added a line that searches for the subelement `<type>` in `<filter>` nodes to support backward compatibility with older XML files (including those in our test suite)
  - Changed the comparison in `load_state_point` to use `settings::n_max_batches` to support continuation runs that use triggers

File: tests/regression_tests/statepoint_restart/test.py [hunk 1]
```diff
@@ -1,11 +1,14 @@
...
 import os
+from pathlib import Path
 
 import openmc
 
 from tests.testing_harness import TestHarness
 from tests.regression_tests import config
+from tests import cdtemp
 
+import pytest
```
Comment (paulromano) [inline tests/regression_tests/statepoint_restart/test.py:11]:
```suggestion
import openmc
import pytest

from tests.testing_harness import TestHarness
from tests.regression_tests import config
from tests import cdtemp
```

**Review by paulromano — CHANGES_REQUESTED** [review]

File: openmc/executor.py [hunk 2]
```diff
@@ -73,8 +74,8 @@ def _process_CLI_arguments(volume=False, geometry_debug=False, particles=None,
         if event_based:
             args.append('-e')
 
-    if isinstance(restart_file, str):
-        args += ['-r', restart_file]
+    if isinstance(restart_file, PathLike.__args__):
```
[the remark this answers is not in the document]
Comment (pshriwise) [reply openmc/executor.py:77]: Yeahh, I agree. I was hoping to be able to make it `openmc.PathLike`, but while this was added in [PEP604](https://peps.python.org/pep-0604/) it seems that isn't available in versions before 3.10.

The nice thing about this approach is that any other types we might include in that `typing.Union` would automatically get checked here, but I don't see that happening anytime soon. In short, I'll be more explicit :)

File: tests/regression_tests/statepoint_restart/test.py [hunk 3]
```diff
@@ -42,7 +43,7 @@ def update_results(self):
 
     def _run_openmc_restart(self):
         # Get the name of the statepoint file.
-        statepoint = glob.glob(os.path.join(os.getcwd(), self._restart_sp))
+        statepoint = list(Path(os.getcwd()).glob(self._restart_sp))
```
Comment (paulromano) [inline tests/regression_tests/statepoint_restart/test.py:46]:
```suggestion
        statepoint = list(Path.cwd().glob(self._restart_sp))
```

**Review by paulromano — CHANGES_REQUESTED** [review]
One more little pathlib trick for ya:

File: tests/regression_tests/statepoint_restart/test.py (hunk 3 above)
Comment (pshriwise) [reply tests/regression_tests/statepoint_restart/test.py:46]: Ooooooh neat!

**Pull request merged**